connector/saml/saml.go

```diff
@@ -0,0 +1,300 @@
+	// support refresh tokens?
+	if s.OfflineAccess {
+		return "", "", fmt.Errorf("SAML does not support offline access")
+	}
+
+	r := &authnRequest{
+		ProtocolBinding: bindingPOST,
+		ID:              "_" + uuidv4(),
+		IssueInstant:    xmlTime(p.now()),
+		Destination:     p.ssoURL,
+		Issuer: &issuer{
+			Issuer: p.issuer,
+		},
+		NameIDPolicy: &nameIDPolicy{
+			AllowCreate: true,
+			Format:      "urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified",
+		},
+		AuthnContextClass: &authnContextClass{
+			Value: "urn:oasis:names:tc:SAML:2.0:ac:classes:PasswordProtectedTransport",
+		},
+	}
+
+	data, err := xml.MarshalIndent(r, "", "  ")
+	if err != nil {
+		return "", "", fmt.Errorf("marshal authn request: %v", err)
+	}
+
+	buff := new(bytes.Buffer)
+	fw, err := flate.NewWriter(buff, flate.DefaultCompression)
+	if err != nil {
+		return "", "", fmt.Errorf("new flate writer: %v", err)
+	}
+	if _, err := fw.Write(data); err != nil {
+		return "", "", fmt.Errorf("compress message: %v", err)
+	}
+	if err := fw.Close(); err != nil {
+		return "", "", fmt.Errorf("flush message: %v", err)
+	}
+
+	return p.ssoURL, base64.StdEncoding.EncodeToString(buff.Bytes()), nil
+}
+
+func (p *provider) HandlePOST(s connector.Scopes, samlResponse string) (ident connector.Identity, err error) {
+	rawResp, err := base64.StdEncoding.DecodeString(samlResponse)
+	if err != nil {
+		return ident, fmt.Errorf("decode response: %v", err)
+	}
+	if p.validator != nil {
+		if rawResp, err = verify(p.validator, rawResp); err != nil {
+			return ident, fmt.Errorf("verify signature: %v", err)
+		}
+	}
+
+	var resp response
+	if err := xml.Unmarshal(rawResp, &resp); err != nil {
+		return ident, fmt.Errorf("unmarshal response: %v", err)
+	}
+
+	if resp.Destination != "" && resp.Destination != p.redirectURI {
+		return ident, fmt.Errorf("expected destination %q got %q", p.redirectURI, resp.Destination)
+
+	}
+
+	assertion := resp.Assertion
+	if assertion == nil {
+		return ident, fmt.Errorf("response did not contain an assertion")
+	}
+	subject := assertion.Subject
+	if subject == nil {
+		return ident, fmt.Errorf("response did not contain a subject")
+	}
+
+	switch {
+	case subject.NameID != nil:
+		if ident.UserID = subject.NameID.Value; ident.UserID == "" {
+			return ident, fmt.Errorf("NameID element does not contain a value")
+		}
+	default:
+		return ident, fmt.Errorf("subject does not contain an NameID element")
+	}
+
+	attributes := assertion.AttributeStatement
+	if attributes == nil {
+		return ident, fmt.Errorf("response did not contain a AttributeStatement")
+	}
+
+	if ident.Email = attributes.get(p.emailAttr); ident.Email == "" {
+		return ident, fmt.Errorf("no attribute with name %q", p.emailAttr)
+	}
+	ident.EmailVerified = true
+
+	if ident.Username = attributes.get(p.usernameAttr); ident.Username == "" {
+		return ident, fmt.Errorf("no attribute with name %q", p.usernameAttr)
+	}
+
+	if s.Groups && p.groupsAttr != "" {
+		groups, ok := attributes.all(p.groupsAttr)
```

Comment:
Maybe add a groupsDelimiter config for IDPs that return groups as comma-separated